mesonpy/_wheelfile.py

```diff
@@ -0,0 +1,99 @@
+# SPDX-License-Identifier: MIT
+
+import base64
+import csv
+import hashlib
+import io
+import os
+import re
+import stat
+import time
+import zipfile
+
+from types import TracebackType
+from typing import List, Optional, Tuple, Type, Union
+
+
+Path = Union[str, os.PathLike]
+
+
+MIN_TIMESTAMP = 315532800  # 1980-01-01 00:00:00 UTC
+WHEEL_FILENAME_REGEX = re.compile(r'^(?P<name>[^-]+)-(?P<version>[^-]+)(:?-(?P<build>[^-]+))?-(?P<tag>[^-]+-[^-]+-[^-]+).whl$')
+
+
+def _b64encode(data: bytes) -> bytes:
+    return base64.urlsafe_b64encode(data).rstrip(b'=')
+
+
+class WheelFile:
+    """Implement the wheel package binary distribution format.
+
+    https://packaging.python.org/en/latest/specifications/binary-distribution-format/
+    """
+    def __new__(cls, filename: Path, mode: str = 'r', compression: int = zipfile.ZIP_DEFLATED) -> 'WheelFile':
+        if mode == 'w':
+            return super().__new__(WheelFileWriter)
+        raise NotImplementedError
```

Comment:
I decided for this indirection to keep open the possibility for implementing a `WheelFileReader` too. I haven't investigated the possibility of modifying a wheel file in place. Maybe all this is over engineered and can be simplified.